### PR TITLE
channel neither has values in nor be closed .

### PR DESCRIPTION
**Pull request by wayilau**
This PR is a 
 - [ ] Feature Implementation
 - [x] Bug Fix
 - [ ] Documentation

**What this PR does / why we need it**:
channel never closed when create a new one by "make", but also has no values in it.

### CODE REVIEW

File: cmd/apiserver/app/server/run_server.go [hunk 1]
```diff
@@ -35,7 +36,7 @@ func RunServer(opts *ServiceCatalogServerOptions, stopCh <-chan struct{}) error
 	if stopCh == nil {
 		/* the caller of RunServer should generate the stop channel
 		if there is a need to stop the API server */
-		stopCh = make(chan struct{})
+		stopCh = server.SetupSignalHandler()
```
Comment (MHBauer) [inline cmd/apiserver/app/server/run_server.go]: should check the callers of this and make sure they're supplying a channel. 

My guess is that there is the cli/cmd , and tests.

We should probably crash if this isn't set, but I haven't looked close enough to know.

Comment (wayilau) [reply cmd/apiserver/app/server/run_server.go]: i will have a look about this. thank you for your review @MHBauer

Comment (jberkhahn) [reply cmd/apiserver/app/server/run_server.go]: Have you looked into this? what happens when it isn't set?

Comment (wayilau) [reply cmd/apiserver/app/server/run_server.go]: sorry, i was busy these days. i Will have a look it today or tomorrow.

Comment (wayilau) [reply cmd/apiserver/app/server/run_server.go]: @jberkhahn the caller give a channel actually, which created by server.SetupSignalHandler(). 
but if stopCh == nil,  just make one can not be closed at all. so i give the code above.

Comment (MHBauer) [reply cmd/apiserver/app/server/run_server.go]: from that description, I think this should panic and crash with an informative message.

Comment (wayilau) [reply cmd/apiserver/app/server/run_server.go]: @MHBauer i modify the code to "panic", also give a message when the stop channel was not set. Please help review about this when you are idle.

File: cmd/apiserver/app/server/run_server.go [hunk 2]
```diff
@@ -35,7 +35,8 @@ func RunServer(opts *ServiceCatalogServerOptions, stopCh <-chan struct{}) error
 	if stopCh == nil {
 		/* the caller of RunServer should generate the stop channel
 		if there is a need to stop the API server */
-		stopCh = make(chan struct{})
+		// stopCh = server.SetupSignalHandler()
```
Comment (jberkhahn) [inline cmd/apiserver/app/server/run_server.go]: should remove the commented line

File: cmd/apiserver/app/server/run_server.go [hunk 3]
```diff
@@ -35,7 +35,8 @@ func RunServer(opts *ServiceCatalogServerOptions, stopCh <-chan struct{}) error
...
+		panic("stop channel was not set when starting the api server")
```
Comment (jberkhahn) [inline cmd/apiserver/app/server/run_server.go]: also, i feel like this should be tested

Comment (wayilau) [reply cmd/apiserver/app/server/run_server.go]: Yes, i've change the code. when RespectsStopCh=true, stopCh is nil,, ,error return, so os.Exit(1) will be called at hyperkube.Run. also  i add the tests

**Comment by wayilau** [comment]
@jberkhahn

File: cmd/apiserver/app/server/run_server.go [hunk 4]
```diff
@@ -33,9 +33,8 @@ import (
 // RunServer runs an API server with configuration according to opts
 func RunServer(opts *ServiceCatalogServerOptions, stopCh <-chan struct{}) error {
 	if stopCh == nil {
-		/* the caller of RunServer should generate the stop channel
-		if there is a need to stop the API server */
-		stopCh = make(chan struct{})
+		/* whe RespectsStopCh: true and the stop channel is nil, error returned*/
```
Comment (jberkhahn) [inline cmd/apiserver/app/server/run_server.go]: I'm not really sure what this comment is trying to say. "Return an error if stopCh is nil"? If that's what it's trying to say, maybe it should be on like 34, above the branch?

File: cmd/apiserver/app/server/run_server.go [hunk 5]
```diff
@@ -33,9 +33,8 @@ import (
...
+		return fmt.Errorf("%v", "the stop channel can not be nil")
```
Comment (MHBauer) [inline cmd/apiserver/app/server/run_server.go]: An error is good enough, but I still think this should be a panic. it's unreachable in normal case. If someone is going to library-ize this code and use it in their own apiserver, they need to know to set it.

File: cmd/apiserver/app/server/run_server.go [hunk 6]
```diff
@@ -33,9 +33,8 @@ import (
 // RunServer runs an API server with configuration according to opts
 func RunServer(opts *ServiceCatalogServerOptions, stopCh <-chan struct{}) error {
 	if stopCh == nil {
-		/* the caller of RunServer should generate the stop channel
-		if there is a need to stop the API server */
```
Comment (MHBauer) [inline cmd/apiserver/app/server/run_server.go]: original comment is better here.

File: cmd/apiserver/app/server/run_server.go (hunk 5 above)
Comment (wayilau) [reply cmd/apiserver/app/server/run_server.go]: yes, i will change the code.

Comment (wayilau) [reply cmd/apiserver/app/server/run_server.go]: @MHBauer i've changed the code comment as it is.

File: cmd/apiserver/app/server/run_server.go (hunk 2 above)
Comment (wayilau) [reply cmd/apiserver/app/server/run_server.go]: @jberkhahn changed.

**Comment by wayilau** [comment]
@MHBauer @jberkhahn 
I've changed the comments as it is. also the panic code when the stopCh not set.

**Comment by MHBauer** [comment]
I'm fine with this
/hold cancel

**Pull request merged**